core/src/main/java/com/sequenceiq/cloudbreak/rotation/context/provider/CMUserRotationContextProvider.java

```diff
@@ -0,0 +1,73 @@
+package com.sequenceiq.cloudbreak.rotation.context.provider;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import com.google.common.collect.Maps;
+import com.sequenceiq.cloudbreak.dto.StackDto;
+import com.sequenceiq.cloudbreak.rotation.context.CMUserRotationContext;
+import com.sequenceiq.cloudbreak.rotation.secret.RotationContext;
+import com.sequenceiq.cloudbreak.rotation.secret.RotationContextProvider;
+import com.sequenceiq.cloudbreak.rotation.secret.SecretLocationType;
+import com.sequenceiq.cloudbreak.service.secret.domain.Secret;
+import com.sequenceiq.cloudbreak.service.secret.service.SecretService;
+import com.sequenceiq.cloudbreak.service.secret.service.rotation.context.VaultRotationContext;
+import com.sequenceiq.cloudbreak.service.stack.StackDtoService;
+import com.sequenceiq.cloudbreak.util.PasswordUtil;
+import com.sequenceiq.cloudbreak.view.ClusterView;
+
+public abstract class CMUserRotationContextProvider implements RotationContextProvider {
+
+    private static final String DATETIMEFORMAT = "ddMMyyHHmm";
+
+    @Inject
+    private StackDtoService stackService;
+
+    @Inject
+    private SecretService secretService;
+
+    @Override
+    public Map<SecretLocationType, RotationContext> getContexts(String resource) {
+        Map<SecretLocationType, RotationContext> result = Maps.newHashMap();
+        StackDto stack = stackService.getByCrn(resource);
+        ClusterView cluster = stack.getCluster();
+
+        Secret userSecret = getUserSecret(cluster);
+        Secret passwordSecret = getPasswordSecret(cluster);
+        Secret clientUserSecret = getClientUserSecret(cluster);
+        Secret clientPasswordSecret = getClientPasswordSecret(cluster);
+        String newUser = getUserPrefix() + new SimpleDateFormat(DATETIMEFORMAT).format(new Date());
+        String newPassword = PasswordUtil.generatePassword();
+        VaultRotationContext vaultRotationContext = VaultRotationContext.builder()
```

Comment:
It's hard to see if this will be correct without the top down implementation of the entire secret rotation.

What I don't understand is that how the context providers and the rotation executor methods will be used in terms of flow steps?

Let's say that we will have the following flow steps:

1. collect contexts from providers,
2. rotate secrets by executors,
3. do the repair,
4. finalize or rollback.

The problem here is that where do you store the secrets between step 1-2. You can't store it for example in the flow context because that is not secure.

Let's say that we will have the following flow steps as an alternative:

1. collect contexts from providers AND rotate secrets by executors,
3. do the repair,
4. finalize or rollback.

Well than if a pod restart happens at the end of the step 1 and after the restart it will face couple of problems:

- it will generate new secrets again and what's worse is that if the vault rotation executor doesn't handle it properly than the original secrets will be lost,
- you can't just check if a new secret was already generated and not put it in the context for vault because than the rollback and finalize steps will have no information to deal with them,

The only option that I see for vault is to check if the secret was already rotated and if so than put in the context the already generated secret and when the vault rotation executor comes it should handle it properly.

Reply:
This is possible that an entirely different flow was imagined here but I don't see how the above issues will be handled at the moment.

Reply:
you are correct, probably we should do the vault rotation first and storing only the path in the context. 
but let's discuss it in the meeting to have a common agreement.